### PR TITLE
actual jooat implementation instead of hooking rage::atStringHash

### PR DESCRIPTION
**Pull request by 0x-FADED**
- why hook `rage::atStringHash(const char*,uint)` when we can have actual jooat implementation
- the code is from here: https://gtaforums.com/topic/879408-c-string-hasher/ from `Fireboyd78`

### CODE REVIEW

File: source/core/joaat.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+namespace SHVDN
+{
+	// implementation of rage::atStringHash(char const*, unsigned int)
+	// look at the pattern for more info E8 ?? ?? ?? ?? 89 03 4C 8B 4D 77
+	public static class atStringHash
```
Comment (kagikn) [inline source/core/joaat.cs:5]: We name classes using Pascal casing so, this should be `AtStringHash`.

File: source/core/joaat.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+			0xBB, 0xBC, 0xBD, 0xBE, 0xBF, 0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5,
+			0xC6, 0xC7, 0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF, 0xD0,
+			0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6, 0xD7, 0xD8, 0xD9, 0xDA, 0xDB,
+			0xDC, 0xDD, 0xDE, 0xDF, 0xE0, 0xE1, 0xE2, 0xE3, 0xE4, 0xE5, 0xE6,
+			0xE7, 0xE8, 0xE9, 0xEA, 0xEB, 0xEC, 0xED, 0xEE, 0xEF, 0xF0, 0xF1,
+			0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB, 0xFC,
+			0xFD, 0xFE, 0xFF
+		};
+
+		private static uint atPartialStringHash(string str, uint initialHash)
```
Comment (kagikn) [inline source/core/joaat.cs:36]: We also name methods using Pascal casing.

**Review by kagikn — CHANGES_REQUESTED** [review]
Sorry, but this PR misses the edge case for double quotes. If the first byte is `0x22`, the function skips the first double quote and hash the string until another double quote or a null character is found. See #501 and search the exe for `48 8B 0B 33 D2 E8 ? ? ? ? 89 03` for more info.
I admit we didn't have to call `rage::atStringHash` since the calculation won't be changed in some game update (no changes between b372 and b2802 at least), though.

In terms of my comments for naming convention for C#, you might want to see [coding conventions in C# documentation](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions). For naming conventions, I think @crosire should review this PR as well.

**Review by kagikn — APPROVED** [review]
Fair enough. If I find any occurrences with strings whose first string is a double quote, I'll add code for the edge case. I haven't seen that occurrences actually.

**Comment by crosire** [comment]
I'm nore sure I see the benefit? I guess we save us a string copy and pin. But it was kinda nice to just have the game do the heavy lifting and not worry about the implementation details.

**Review by kagikn — CHANGES_REQUESTED** [review]
I forgot to say this, but strings (which is UTF-16 in .NET) should be encoded to UTF-8 just like before. GTA V uses UTF-8 and ASCII strings, but definitely not UTF-16 strings.
The current implementation throws exception if the string contains non-ASCII characters (`IndexOutOfRangeException`).

**Comment by kagikn** [comment]
@crosire I'm not sure if you are saying we should not worry about the implementation details of the hashing function, but I can say this PR will improve performance (about 2x faster than before). Besides, I don't think they don't change how to calculate joaat hashes because even the equivalent function in GTA IV calculates in the same way (the function is at `0x5B1C30` in v1.0.7.0).

```
public static string RandomString(int length)
{
    const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_@";
    return new string(Enumerable.Repeat(chars, length)
                                .Select(s => s[randomInstForRandomStr.Next(s.Length)]).ToArray());
}

public static void MeasurePerformance()
{
    var randomStrs = new List<string>(10000000);
    for (int i = 0; i < 1000000; i++)
    {
        randomStrs.Add(RandomString(12));
    }

    var stopwatch = new Stopwatch();
    stopwatch.Start();
    foreach (var randomStr in randomStrs)
    {
        Game.GenerateHashOld(randomStr); // call rage::atStringHash via a delegate obtained via Marshal.GetDelegateForFunctionPointer (before this PR)
    }
    stopwatch.Stop();
    var oldElapsedTime = stopwatch.ElapsedMilliseconds;

    stopwatch.Reset();
    stopwatch.Start();
    foreach (var randomStr in randomStrs)
    {
        Game.GenerateHash(randomStr); // calculate the hash value by directly executing C# code (this PR)
    }
    stopwatch.Stop();
    var newElapsedTime = stopwatch.ElapsedMilliseconds;
}
```
In this code, `oldElapsedTime` will be twice as many as `newElapsedTime`. I guess calling a delegate consumes much more time when the delegate is obtained via `Marshal.GetDelegateForFunctionPointer()` than we would think.

**Comment by crosire** [comment]
Fair. From an implementation point of view I'd just keep `NativeMemory.GetHashKey(string key) though and simply replace its definition, since the fancy concatenation overloads aren't used anyway:
```c#
static readonly byte[] LookupTable = { ... };

public static uint GetHashKey(string key)
{
	var bytes = Encoding.UTF8.GetBytes(key);
	uint hash = 0;
	for (int i = 0; i < str.Length; i++)
	{
		hash += LookupTable[bytes[i]];
		hash += (hash << 10);
		hash ^= (hash >> 6);
	}
	hash += (hash << 3);
	hash ^= (hash >> 11);
	hash += (hash << 15);
	return hash;
}
```

**Comment by Sardelka9515** [comment]
> I guess calling a delegate consumes much more time when the delegate is obtained via `Marshal.GetDelegateForFunctionPointer()` than we would think.

A bit irrelevant here, but i just figured that we can use the [function pointer syntax](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers), just need to specify the calling convention (presumably Stdcall), maybe with the SuppressGCTransition attribute. Haven't tested yet, but it should work and be much faster.

**Review by kagikn — APPROVED** [review]
Fine by me regarding how to hash strings.

**Comment by kagikn** [comment]
Sorry, but conflicts occur since I merged #1136.

**Comment by crosire** [comment]
Would also prefer to ditch the new joaat.cs file and move the function definition back into the behemoth that is the `NativeMemory` class.

**Pull request closed without merging**